pkg/plugins/policies/meshhttproute/plugin/v1alpha1/listeners.go

```diff
@@ -0,0 +1,156 @@
+
+			listenerBuilder = listenerBuilder.Configure(envoy_listeners.FilterChain(filterChainBuilder))
+		}
+
+		listener, err := listenerBuilder.Build()
+		if err != nil {
+			return nil, err
+		}
+		resources.Add(&core_xds.Resource{
+			Name:     listener.GetName(),
+			Origin:   generator.OriginOutbound,
+			Resource: listener,
+		})
+	}
+
+	return resources, nil
+}
+
+// Whenever `split` is specified in the TrafficRoute which has more than kuma.io/service tag
+// We generate a separate Envoy cluster with _X_ suffix. SplitCounter ensures that we have different X for every split in one Dataplane
+// Each split is distinct for the whole Dataplane so we can avoid accidental cluster overrides.
+type splitCounter struct {
+	counter int
+}
+
+func (s *splitCounter) getAndIncrement() int {
+	counter := s.counter
+	s.counter++
+	return counter
+}
+
+func makeClusters(
+	proxy *core_xds.Proxy,
+	clusterCache map[string]string,
+	splitCounter *splitCounter,
+	refs []api.BackendRef,
+) []envoy.Cluster {
+	var clustersInternal, clustersExternal []envoy.Cluster
+
+	for _, ref := range refs {
+		switch ref.Kind {
+		case common_api.MeshService, common_api.MeshServiceSubset:
+		default:
+			continue
+		}
+
+		service := ref.Name
+		if ref.Weight == 0 {
+			continue
+		}
+
+		name := service
+
+		if len(ref.Tags) > 0 {
+			name = envoy_names.GetSplitClusterName(service, splitCounter.getAndIncrement())
+		}
+
+		if mesh, ok := ref.Tags[mesh_proto.MeshTag]; ok {
+			// The name should be distinct to the service & mesh combination
+			name = fmt.Sprintf("%s_%s", name, mesh)
+		}
+
+		// We assume that all the targets are either ExternalServices or not
+		// therefore we check only the first one
+		var isExternalService bool
+		if endpoints := proxy.Routing.OutboundTargets[service]; len(endpoints) > 0 {
+			isExternalService = endpoints[0].IsExternalService()
+		}
+		if endpoints := proxy.Routing.ExternalServiceOutboundTargets[service]; len(endpoints) > 0 {
+			isExternalService = true
+		}
+
+		allTags := envoy_tags.Tags(ref.Tags).WithTags(mesh_proto.ServiceTag, ref.Name)
+		cluster := envoy_common.NewCluster(
+			envoy_common.WithService(service),
+			envoy_common.WithName(name),
+			envoy_common.WithWeight(uint32(ref.Weight)),
+			// The mesh tag is set here if this destination is generated
+			// from a MeshGateway virtual outbound and is not part of the
+			// service tags
+			envoy_common.WithTags(allTags.WithoutTags(mesh_proto.MeshTag)),
+			envoy_common.WithExternalService(isExternalService),
+		)
+
+		if mesh, ok := ref.Tags[mesh_proto.MeshTag]; ok {
+			cluster.SetMesh(mesh)
+		}
+
+		if name, ok := clusterCache[allTags.String()]; ok {
+			cluster.SetName(name)
+		} else {
+			clusterCache[allTags.String()] = cluster.Name()
+		}
+
+		if isExternalService {
+			clustersExternal = append(clustersExternal, cluster)
+		} else {
+			clustersInternal = append(clustersInternal, cluster)
+		}
+	}
+
```

Comment:
I think it can be removed. It's used in `OutboundProxyGenerator` for conditionally applying rate limits but this will have to be done in the `MeshRateLimit` plugin.